sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java

```diff
@@ -841,20 +853,150 @@ private void ensureToNotCalledYet() {
+          SerializableFunction<ValueInSingleWindow<T>, TableDestination> tableFunction) {
+        this.tableFunction = tableFunction;
+      }
+
+      @Override
+      public TableDestination getDestination(ValueInSingleWindow<T> element) {
+        return tableFunction.apply(element);
+      }
+
+      @Override
+      public TableSchema getSchema(TableDestination destination) {
+        return null;
+      }
+
+      @Override
+      public TableDestination getTable(TableDestination destination) {
+        return destination;
+      }
+
+      @Override
+      public Coder<TableDestination> getDestinationCoder() {
+        return TableDestinationCoder.of();
+      }
+    }
+
+    /**
+     * Delegates all calls to an inner instance of {@link DynamicDestinations}. This allows
+     * subclasses to modify another instance of {@link DynamicDestinations} by subclassing and
+     * overriding just the methods they want to alter.
+     */
+    static class DelegatingDynamicDestinations<T, DestinationT>
+        extends DynamicDestinations<T, DestinationT> {
+      private final DynamicDestinations<T, DestinationT> inner;
+      DelegatingDynamicDestinations(DynamicDestinations<T, DestinationT> inner) {
+        this.inner = inner;
+      }
+      @Override
+      public DestinationT getDestination(ValueInSingleWindow<T> element) {
+        return inner.getDestination(element);
+      }
+
+      @Override
+      public TableSchema getSchema(DestinationT destination) {
+        return inner.getSchema(destination);
+      }
+
+      @Override
+      public TableDestination getTable(DestinationT destination) {
+        return inner.getTable(destination);
+      }
+
+      @Override
+      public Coder<DestinationT> getDestinationCoder() {
+        return inner.getDestinationCoder();
+      }
+    }
+    /**
+     * Returns the same schema for every table.
+     */
+    static class ConstantSchemaDestinations<T>
+        extends DelegatingDynamicDestinations<T, TableDestination> {
+      private final @Nullable ValueProvider<String> jsonSchema;
+
+      ConstantSchemaDestinations(DynamicDestinations<T, TableDestination> inner,
+                                 ValueProvider<String> jsonSchema) {
+        super(inner);
+        this.jsonSchema = jsonSchema;
+      }
+
+      @Override
+      public TableSchema getSchema(TableDestination destination) {
+        return BigQueryHelpers.fromJsonString(jsonSchema.get(), TableSchema.class);
+      }
+    }
+
+    /**
+     * Takes in a side input mapping tablespec to json table schema, and always returns the
+     * matching schema from the side input.
+     */
+    static class SchemaFromViewDestinations<T>
+        extends DelegatingDynamicDestinations<T, TableDestination> {
+      SchemaFromViewDestinations(DynamicDestinations<T, TableDestination> inner,
+                                 PCollectionView<Map<String, String>> schemaView) {
+        super(inner);
+        withSideInput(schemaView);
```

Comment:
The method withSideInput should be renamed to setSideInput - `with` is traditionally non-mutating.